### ISSUE TITLE
UNASSIGNED-CoreValidation-Shader-InconsistentSpirv reported when using sampler2DRect in GLSL

### ISSUE THREAD

**Issue by EuwTech** [KhronosGroup/Vulkan-Docs#1582]
When loading a SPIR-V module compiled with a sampler2DRect either as a split texture2DRect and sampler, the following error is generated:

` UNASSIGNED-CoreValidation-Shader-InconsistentSpirv -  SPIR-V module not valid: Capability SampledRect is not allowed by Vulkan 1.2 specification (or requires extension)`

[See this khronos Vulkan forum thread for more info.](https://community.khronos.org/t/unnormalized-texture-coordinate-reads-with-vksampler/107543/7)

**Comment by gnl21** [comment]
This was an intentional change from GL to Vulkan (to use "normal" texture types for unnormalised coordinates, rather than the specific "Rect" type). The validation error message could be more helpful, and we'll improve that, and the state of GLSL makes this confusing. We're proposing to update KHR_vulkan_glsl to disallow "Rect" textures when targeting Vulkan.

I think that the method in the linked thread (using sampler2D, and textureLod with lod 0) is the correct method for doing this in Vulkan.

**Comment by krOoze** [comment]
> I think that the method in the linked thread (using sampler2D, and textureLod with lod 0) is the correct method for doing this in Vulkan.

Feels like awkward workaround. I think ideally `texture()` should just work. If only for OpenGL and Vulkan shader compatibility. It feels this is too basic thing to have to `ifdef` around.

**Comment by gnl21** [comment]
I agree that it's a bit awkward, but from a Vulkan/SPIR-V perspective it makes sense to only allow ExplicitLod lookups, since the ImplicitLod calculations won't work for unnormalised coordinates. That's the same as for vertex shaders, for which the front-end has to take responsibility for not using ImplicitLod where it's not supported.

Unfortunately GLSL doesn't have enough information for front-ends to be able to emit the correct instruction without assistance from the shader author, so `texture` ends up not working as you'd expect. Perhaps it would have been better to add the information to GLSL, but it's not there, and because the restriction on ImplicitLod has always been in the spec drivers are not generating the code that would be required to make `texture` just work.

One alternative solution for GLSL would be for compilers to accept sampler2DRect for Vulkan but compile it down to a 2D texture access with implicit lod disabled. I don't think that was the original intention but it might make for slightly better compatibility with OpenGL. The problem would remain for other types of texture that support unnormalised coordinates though.

**Comment by krOoze** [comment]
Is there any case one would actually **want** to bind unnormalized sampler to non-Rect `sampler2D`? Should perhaps not been allowed in the first place, and been `Rect` all the time. Could just deprecate that use, and warn in the case `sampler2D` is bound to unnormalized sampler. In that case the `textureLOD` is fine, and anyone making new code would just use `Rect`.

**Comment by gnl21** [comment]
Are you suggesting changing GLSL or SPIR-V? I'm not really sure that this is enough of a problem to warrant changing the design of these things.

There are various problems with compatibility with the GL way of doing this, like that the dimensionality is associated with the image but unnormalized coordinates are associated with the sampler, 1D textures have no corresponding rect-like dimension. I agree that it's not perfect, but the alternative design also has issues that would need working out, so is this really a big problem?

**Comment by krOoze** [comment]
Well, I think esthetics is the last priority of Vulkan. But I think it still is a priority. Perhaps not a big problem, but it adds up if there are 100 similar small problems, and it reflects on the general reputation of Vulkan. Since Vulkan is extensible, I also tend to believe there are no small issues in the foundations.

> 1D textures have no corresponding rect-like dimension

Hm, true. Opens the question if normalizeness should be part of the type. I see it is in SPIR-V for constant sampler at least.

> Are you suggesting changing GLSL or SPIR-V?

Suggesting nothing specific. Depends what would be best...

The fix could potentially be trivial and isolated to `GL_KHR_vulkan_glsl` (resp. to shader compilers working in Vulkan semantics).

But alternatively could just allow implicit LOD SPIR-V commands on unnormalized image. I mean the images are currently forced to have only one level. so the implicit level should conceptually work anyway.

**Comment by EuwTech** [comment]
Take a look at this [post on stack overflow](https://stackoverflow.com/questions/65790303/how-to-use-a-vulkan-sampler-with-unnormalized-texture-coordinates-without-trig/68410130#68410130), it is fairly recent (6 months old) and answered by someone with a high rep over there, and they seem to think that unnormalized isn't even possible with GLSL on Vulkan. (I added an answer below it in an attempt to help clear things up)

It's pretty murky figuring out what a valid unnormalized read from GLSL is by just reading the spec so folks tend to go by remarks like the one in this link. So whatever decision is made around this, hopefully the spec and/or validation layer errors can be changed to help point us lost soles in the right direction.

**Comment by EuwTech** [comment]
Link added, and thank you!

**Comment by gnl21** [comment]
No, this slipped through the cracks, thanks for the reminder. I've created a PR here: https://github.com/KhronosGroup/GLSL/pull/224 and will update the next version of the core spec with the changes. Let me know what you think.
